### PR TITLE
HIVE-17608: REPL LOAD should overwrite the data files if exists instead of duplicating it

### PR DESCRIPTION
**Pull request by sankarh**

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 1]
```diff
@@ -1425,7 +1425,66 @@ public NotificationEventResponse apply(@Nullable NotificationEventResponse event
...
+                // Add insert event twice with different event ID to allow apply of both events.
+                NotificationEvent newEvent = new NotificationEvent(event);
+                outEvents.add(newEvent);
+                insertEventId = newEvent.getEventId();
+              }
+            }
+
+            NotificationEvent newEvent = new NotificationEvent(event);
+            if (insertEventId != -1) {
+              insertEventId++;
```
[the remark this answers is not in the document]
Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: How event id becomes 0?  As I use the last insert event ID and duplicate it once with event ID +1 and then increment it for each event which comes afterwards. So, there shouldn't be any out of order case here.

[the remark this answers is not in the document]
Comment (sankarh) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: Also looks like in ReplicationSpec if we set the isReplace to true then we replace_all else we do insert_overwrite which are basically the same thing. so why not remove that variable from replicationspec and always do a replace_all/insert_overwrite by default ?

-- replace_all and overwrite_existing are not same. replace_all will remove all files in the table/partitions and adds only the new ones. Overwrite_existing will replace only those files which has the name matching with new ones.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -3147,6 +3152,10 @@ private static Path mvFile(HiveConf conf, FileSystem sourceFs, Path sourcePath,
     * I'll leave the below loop for now until a better approach is found.
     */
     for (int counter = 1; destFs.exists(destFilePath); counter++) {
+      if (isOverwrite) {
+        destFs.delete(destFilePath, false);
```
[the remark this answers is not in the document]
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: There is a method in rename which takes overwrite as input which I don't want to use as it is deprecated. anyways, overwrite flag also does the same thing by deleting the existing file which I did it in caller.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 3]
```diff
@@ -3147,6 +3152,10 @@ private static Path mvFile(HiveConf conf, FileSystem sourceFs, Path sourcePath,
...
+        break;
```
[the remark this answers is not in the document]
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: If I delete the existing destination file, then need not change the name of the destfile path as we shall copy with the same name. That's why break from loop after deleting it.

[the remark this answers is not in the document]
Comment (anishek) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: the loop however will use the file with the last counter number before your change but now will have original file name and first file found with the counter to be deleted. is that correct for all use cases ?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (anishek) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: yeh missed the != -1 the above works.

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sankarh) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java]: Yes, if overwrite flag is set, we care only about the original file name. If file exist with original name, we just need to delete and proceed to copy the new one. 
The loop is there for the KEEP_EXISTING/overwrite=false case where need to retain the existing file and hence change the destination file name using counter.

**Pull request closed without merging**